File: recursion/lists.js
const util = require('node:util');

const EMPTY_LIST = null;

class Node {
  /**
   * Helper function to construct a linked list from an array of values
   */
  static fromArray(array) {
    if (array.length === 0) {
      return EMPTY_LIST;
    }

    let [first, ...rest] = array;

    return prepend(first, Node.fromArray(rest));
  }

  constructor(value, next = EMPTY_LIST) {
    this.value = value;
    this.next = next;
  }

  toString() {
    // Use an external function so we can display empty lists as "()"
    // rather than "null"
    return listToString(this);
  }

  // Changes how instances are displayed in the Node console/repl
  // This is like Python's __repr__ magic method and is Node-specific
  [util.inspect.custom]() {
    return this.toString();
  }

  *[Symbol.iterator]() {
    yield this.value;

    if (!isEmpty(this.next)) {
      yield* this.next;
    }
  }
}

/**
 * Helper function to pretty-print linked lists. Defined outside
 * the Node class so that we can use `null` to represent the empty list.
 */
function listToString(list) {
  if (isEmpty(list)) {
    return '()';
  }

  let [first, rest] = unprepend(list);

  return `${first} -> ${listToString(rest)}`;
}

/**
 * Returns true if list is empty and false otherwise
 */
function isEmpty(list) {
  return list === EMPTY_LIST;
}

/**
 * Returns true if rest of list is empty and false otherwise
 * That is, returns true if list has only one element
 */
function isRestEmpty(list) {
  return isEmpty(list) || isEmpty(list.next);
}

/**
 * Construct a new list by prepending a value to another list
 */
function prepend(val, list) {
  if (isEmpty(list)) {
    return new Node(val);
  }

  return new Node(val, list);
}

/**
 * Returns a pair consisting of the first element of the list and the rest of the list
 */
function unprepend(list) {
  return [list.value, list.next];
}

module.exports = {
  Node,
  prepend,
  unprepend,
  EMPTY_LIST,
  listToString,
  isEmpty,
  isRestEmpty,
};

File: recursion/python/lists.py
EMPTY_LIST = None

class Node:
    # Helper method to generate a linked list from a list
    @staticmethod
    def from_list(lst):
        if len(lst) == 0:
            return EMPTY_LIST

        first, *rest = lst
        return prepend(first, Node.from_list(rest))

    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def __str__(self):
        # Use an external function so we can display empty lists as "()"
        # rather than "None"
        return list_to_string(self)

    def __repr__(self):
        return self.__str__()

def list_to_string(lst):
    if is_empty(lst):
        return '()'

    first, rest = unprepend(lst)
    return f'{first} -> {list_to_string(rest)}'

def is_empty(lst):
    return lst is EMPTY_LIST

def is_rest_empty(lst):
    return is_empty(lst) or is_empty(lst.next)

def prepend(val, lst):
    if is_empty(lst):
        return Node(val)

    return Node(val, lst)

def unprepend(lst):
    return lst.value, lst.next
